### ISSUE TITLE
Identify line numbers in JSON documents of findings

### ISSUE THREAD

**Issue by 0xdabbad00** [duo-labs/parliament#62]
No existing Python libraries for parsing JSON maintain a correlation between the elements and the line numbers. As a result, the "location" of a finding is given as the text of the element, which isn't very precise and can be ambiguous, especially in a hundred line policy.  

Based on a comment from Paul McGuire (https://twitter.com/ptmcguire/status/1216604117826723842), it should be possible to use https://github.com/pyparsing/pyparsing to keep track of the line number.

**Comment by 0xdabbad00** [comment]
Along this same concept, I need to have a `location` element associated with each `Statement` object.  For example with CloudMapper's `access_check`, I identity what principals have certain privileges, which really is just letting me see all of the statements associated with a principal.  I then end up with a Statement that is just `Effect: Allow, Action: *, Resource: *` and I need to try to dig through the `iam-get-account-authorization-details.json` to figure out where this came from.

**Comment by 0xdabbad00** [comment]
I attempted to use pyparsing, but it looks like it is going to be a fair bit of work still.  I did `pip install pyparsing` then copied the contents of https://github.com/pyparsing/pyparsing/blob/master/examples/jsonParser.py to a file.  

Using `print(results)` shows:
```
[['glossary', [['title', 'example glossary'], ['GlossDiv', [['title', 'S'], ['GlossList', [['ID', 'SGML'], ['SortAs', 'SGML'], ['GlossTerm', 'Standard Generalized Markup Language'], ['TrueValue', True], ['FalseValue', False], ['Gravity', -9.8], ['LargestPrimeLessThan100', 97], ['AvogadroNumber', 6.02e+23], ['EvenPrimesGreaterThan2', None], ['PrimesLessThan10', [2, 3, 5, 7]], ['Acronym', 'SGML'], ['Abbrev', 'ISO 8879:1986'], ['GlossDef', 'A meta-markup language, used to create markup languages such as DocBook.'], ['GlossSeeAlso', ['GML', 'XML', 'markup']], ['EmptyDict', []], ['EmptyList', [[]]]]]]]]]]
```

You can see everything is converted to an array, so you lose the concept of elements.

Using the code from Paul McQuire's tweet, I used:
```
locns = []
    memberDef.addParseAction(lambda s, l, t: 
                             locns.append((t[0][0], pp.lineno(l, s))))
    results = jsonObject.parseString(testdata)
    for loc in locns:
        print(loc)
```

This prints:
```
('title', 4)
('title', 6)
('ID', 9)
('SortAs', 10)
('GlossTerm', 11)
('TrueValue', 12)
('FalseValue', 13)
('Gravity', 14)
('LargestPrimeLessThan100', 15)
('AvogadroNumber', 16)
('EvenPrimesGreaterThan2', 17)
('PrimesLessThan10', 18)
('Acronym', 19)
('Abbrev', 20)
('GlossDef', 21)
('GlossSeeAlso', 22)
('EmptyDict', 23)
('EmptyList', 24)
('GlossList', 7)
('GlossDiv', 5)
('glossary', 3)
```

So pyparsing can figure out the line numbers, but the ordering is changed, and this is just an array of elements, so it doesn't maintain the structure.

My conclusion is that while pyparsing may allow this to happen, it would require learning how that library works and doing a lot more work to it.

Keep in mind also that Parliament assume it is just working on a big structure of maps and arrays.  So you would have to change a ton of code.  I think all we really need is the line number for the statements, so as long as we can track the location to that point, we would be ok, and we wouldn't need to change so much code that way.

**Comment by 0xdabbad00** [comment]
Looks like I can use https://pypi.org/project/json-cfg/
It provides the function `jsoncfg.node_location(node)` containing `NodeLocation(line, column)`

**Comment by 0xdabbad00** [comment]
I started the conversion and found the json-cfg library is noticeably slower.  Here are the times from running `time bin/parliament` which doesn't do anything except load the `iam_definition.json` file and then prints an error message about the parameters needed:

`0m0.212s` - Using json
`0m3.940s` - Using jsoncfg

The `iam_definition.json` file is 3.5MB and 120K lines, so I'm going to try to get the best of both worlds, by using `json` for the IAM definition file, and `jsoncfg` for the user provided file.  I do think having the line numbers is worth the performance delay, but there is not going to be a way to choose performance or line numbers, so I'm going all in with this change.

**Comment by 0xdabbad00** [comment]
It looks like the object I end up with is a tuple containing the text of the current element, and then an object for the member that contains the location, which means the location I present is always going to be for the member, which is annoying.

**Comment by 0xdabbad00** [comment]
This is resolved via #139
